README
HAVAL Version 1.1
=================

Copyright (c) 2003 Calyptix Security Corporation
Copyright (c) 2020 Mike Gelfand
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above
    copyright notice, this list of conditions and the following
    disclaimer in the documentation and/or other materials provided
    with the distribution.

 3. Neither the name of Calyptix Security Corporation nor the
    names of its contributors may be used to endorse or promote
    products derived from this software without specific prior
    written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

-------------------------------------------------------------------

Introduction
============

Files
=====

README         this file
haval.h        interface to the HAVAL (V.1) hashing library
haval.c        the HAVAL (V.1) hashing library
havaltest.c    test program for the library
haval.cert     HAVAL certification data set
pi.frac        a test data file for HAVAL


ChangeLog
2020-11-21  Mike Gelfand  <mikedld@mikedld.com>

    * Switch from separate `from_*` methods to an overloaded `hash` method.
    * Add new `end_to` method to store result in a user-provided buffer.
    * Expose library version via havalver.h file.
    * Add an (optional) Qt convenience wrapper with Qt 5 and Qt 6 support.
    * Support specifying components when using `find_package` with CMake;
      currently 3 components are provided: "core" (STL-based library),
      "qt" (Qt-based wrapper), and "app" (CLI tool).

2020-10-24  Mike Gelfand  <mikedld@mikedld.com>

    * Rewrite the library in C++ and make it header-only, allowing to
      instantiate the template with different pass count and fingerprint
      length values within the same program without much fuss.
    * Move certification data checks code from the app into a unit test.
    * Allow changing app parameters without recompilation via environment
      variables: HAVAL_PASS (defaults to 3), HAVAL_FPTLEN (256),
      HAVAL_NUMBER_OF_BLOCKS (5000), HAVAL_BLOCK_SIZE (5000).
    * Switch to CMake build system.

2003-01-21  Lawrence Teo  <lteo@calyptix.com>

    * Makefile.am (EXTRA_DIST): Distribute haval-paper.pdf and not
      haval.ps.

2003-01-20  Lawrence Teo  <lteo@calyptix.com>

    * All source files: Added/updated preambles to reflect new license.

2003-01-19  Lawrence Teo  <lteo@calyptix.com>

    * Repackaged for Calyptix Security Corporation.
    * Relicensed under the BSD license.
    * Created this ChangeLog.
    * TODO: Added. To-do list.
    * COPYING.BSD: Added. The BSD license.
    * INSTALL: Added. Build and installation instructions.
    * Used GNU autoconf and automake to facilitate project management.
    * Tested compilation with Debian GNU/Linux 3.0/x86, Red Hat
      Linux 8.0/x86, OpenBSD 3.0/x86, and Solaris 8/SPARC,
    * configure.in: Added.
    * Makefile.am: Added. Forced -Wall switch to gcc, so that all
      warnings are displayed.
    * acinclude.m4: Added. Self-defined m4 macros.
    * bootstrap: Added. Bootstraps the project by invoking the
      autotools.
    * configure.in: Enabled the user to specify number of passes with
      the --with-num-passes switch to the configure switch, e.g.
      ./configure --with-num-passes=3
    * configure.in: Enabled fingerprint length to be specified
      using --with-fpt-len.
    * configure.in: Enabled number of test blocks to be specified
      using --with-test-blocks.
    * configure.in: Enabled test block size to be specified using
      --with-block-size.
    * configure.in: Preliminary check for endianness, so that user
      does not have to define/undefine LITTLE_ENDIAN manually. Still
      needs more work, since LITTLE_ENDIAN exists in
      /usr/include/endian.h on Linux systems.
    * configure.in: Added the --enable-extra-optimization switch,
      which will get gcc to perform extra optimization (for i686 only).
    * havaltest.c: Added Id tag to facilitate binary identification.
    * havaltest.c: When producing the hash of a file (say, file.txt),
      print HAVAL(file.txt) instead of HAVAL(File file.txt) to make
      it more consistent with other hashing programs.
    * havaltest.c (usage): New. Usage is now printed in a more uniform
      manner like other UNIX software.
    * havaltest.c: Include config.h instead of havalapp.h.
    * haval.c: Added Id tag to facilitate binary identification.
    * haval.c: s/VERSION/HAVAL_VERSION/ to avoid conflicts and
      confusion with package version generated by autotools.
    * haval.c (haval_stdin): Fixed bug where hash of stdin produces
      raw characters. Now produces regular hexadecimal output.
    * haval.c (haval_stdin): Got rid of compile warning because of the
      use of assignment as a truth value without parentheses.
    * haval.c (haval_file): Got rid of compile warning because of the
      use of assignment as a truth value without parentheses.
    * havalapp.h: Moving towards deprecation, once config.h is
      stabilized.

1997-04-01  Yuliang Zheng  <yuliang@calyptix.com>

    * haval.c: Corrected the first byte in padding[128]
      (from 0x80 to 0x01).
    * Thanks go to Paulo Barreto (pbarreto@uninet.com.br) who
      detected the padding error while implementing HAVAL
      independently.

1996-04-01  Yuliang Zheng  <yuliang@calyptix.com>

    * haval.c: Added #include <string.h>
    * haval.c: Deleted #include <memory.h>
    * haval.c: Appended 'L' to all 32-bit constants.
    * Many thanks to Ross Williams (ross@rocksoft.com) for his
      invaluable comments on the previous version of the code.
    * havaltest.c: Now we use CPU time, instead of elapsed time to
      measure the speed of HAVAL.

1993-06-01  Yuliang Zheng  <yuliang@calyptix.com>

	First release.


src/havalapp.cpp
// Copyright (c) 2020, Mike Gelfand
// Copyright (c) 2003, Calyptix Security Corporation
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this
//    list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its
//    contributors may be used to endorse or promote products derived from
//    this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include "haval.hpp"

#include <fstream>
#include <iomanip>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>

#include <time.h>

namespace
{

unsigned int get_env_uint(const char* key, unsigned int default_value)
{
    const char* str_value = std::getenv(key);
    int int_value = str_value != nullptr ? std::atoi(str_value) : 0;
    return int_value > 0 ? static_cast<unsigned int>(int_value) : default_value;
}

// test the speed of HAVAL
template<unsigned int pass_cnt, unsigned int fpt_len>
void haval_speed()
{
    const unsigned int blocks_cnt = get_env_uint("HAVAL_NUMBER_OF_BLOCKS", 5000);
    const unsigned int block_size = get_env_uint("HAVAL_BLOCK_SIZE", 5000);

    std::cout << "Test the speed of HAVAL (PASS = " << pass_cnt << ", FPTLEN = " << fpt_len << " bits)." << std::endl;
    std::cout << "Hashing " << blocks_cnt << " " << block_size << "-byte blocks ..." << std::endl;

    // initialize test block
    const std::unique_ptr<std::uint8_t[]> buff(new std::uint8_t[block_size]);
    for (unsigned int i = 0; i < block_size; i++) {
        buff[i] = static_cast<std::uint8_t>(~0);
    }

    haval::haval<pass_cnt, fpt_len> hasher;
    std::string fingerprint;

    // reset the clock
    clock();

    // hash
    hasher.start();
    for (unsigned int i = 0; i < blocks_cnt; i++) {
        hasher.update(buff.get(), block_size);
    }
    fingerprint = hasher.end();

    // get the number of clocks
    const clock_t clks = clock();
    // get cpu time
    const double cpu_time = static_cast<double>(clks) / static_cast<double>(CLOCKS_PER_SEC);

    if (cpu_time > 0.0) {
        std::cout << "CPU Time = " << std::fixed << std::setprecision(1) << cpu_time << " seconds" << std::endl;
        std::cout << "   Speed = " << std::fixed << std::setprecision(2) << ((blocks_cnt * block_size * 8) / (1.0E6 * cpu_time))
                  << " MBPS (megabits/second)" << std::endl;
    } else {
        std::cout << "not enough blocks !" << std::endl;
    }
}

// test endianity
bool little_endian()
{
    const std::uint8_t str[4] = {'A', 'B', 'C', 'D'};
    const haval::detail::word_t* wp = reinterpret_cast<const haval::detail::word_t*>(str);
    return str[0] == static_cast<std::uint8_t>(*wp & 0xFF);
}

// print a fingerprint in hexadecimal
std::string to_hex(const std::string& fingerprint)
{
    std::ostringstream stream;
    stream << std::hex << std::uppercase << std::setfill('0');
    for (char c : fingerprint) {
        stream << std::setw(2) << int{static_cast<std::uint8_t>(c)};
    }
    return stream.str();
}

// print usage
void usage(unsigned int pass_cnt, unsigned int fpt_len)
{
    std::cerr << "Usage: haval [OPTION] [FILE]..." << std::endl
              << "  or:  haval -m [STRING]" << std::endl
              << "Generates HAVAL hashes." << std::endl
              << "With no FILE, read standard input." << std::endl
              << std::endl
              << "Configured to use " << pass_cnt << " passes and a " << fpt_len << "-bit fingerprint length." << std::endl
              << std::endl
              << "    ?/-?/-h    show help menu" << std::endl
              << "    -e         test endianity" << std::endl
              << "    -m string  hash the given string" << std::endl
              << "    -s         test speed" << std::endl
              << std::endl
              << "Report bugs to <info@calyptix.com>." << std::endl;
}

template<unsigned int pass_cnt, unsigned int fpt_len>
int main_impl(int argc, char* argv[])
{
    using hasher = haval::haval<pass_cnt, fpt_len>;

    if (argc <= 1) {
        // filter
        std::cout << to_hex(hasher::hash(std::cin)) << std::endl;
    }

    for (int i = 1; i < argc; i++) {
        const std::string arg = argv[i];

        if (arg == "?" || arg == "-?" || arg == "-h") {
            // show help info
            usage(pass_cnt, fpt_len);
        } else if (arg.compare(0, 2, "-m") == 0) {
            // hash string
            const std::string data = arg.substr(2);
            std::cout << "HAVAL(" << std::quoted(data) << ") = " << to_hex(hasher::hash(data)) << std::endl;
        } else if (arg == "-s") {
            // test speed
            haval_speed<pass_cnt, fpt_len>();
        } else if (arg == "-e") {
            // test endianity
            if (little_endian()) {
                std::cout << "Your machine is little-endian." << std::endl;
                std::cout << "You may define HAVAL_LITTLE_ENDIAN to speed up processing." << std::endl;
            } else {
                std::cout << "Your machine is NOT little-endian." << std::endl;
                std::cout << "You must NOT define HAVAL_LITTLE_ENDIAN." << std::endl;
            }
        } else {
            // hash file
            std::ifstream f(arg.c_str(), std::ios::in | std::ios::binary);
            if (!f.good()) {
                std::cout << arg << " can not be opened !" << std::endl;
            } else {
                std::cout << "HAVAL(" << arg << ") = " << to_hex(hasher::hash(f)) << std::endl;
            }
        }
    }

    return 0;
}

template<unsigned int pass_cnt>
int main_impl(unsigned int fpt_len, int argc, char* argv[])
{
    switch (fpt_len) {
    case 128:
        return main_impl<pass_cnt, 128>(argc, argv);
    case 160:
        return main_impl<pass_cnt, 160>(argc, argv);
    case 192:
        return main_impl<pass_cnt, 192>(argc, argv);
    case 224:
        return main_impl<pass_cnt, 224>(argc, argv);
    case 256:
    default:
        return main_impl<pass_cnt, 256>(argc, argv);
    }
}

int main_impl(unsigned int pass_cnt, unsigned int fpt_len, int argc, char* argv[])
{
    switch (pass_cnt) {
    case 3:
    default:
        return main_impl<3>(fpt_len, argc, argv);
    case 4:
        return main_impl<4>(fpt_len, argc, argv);
    case 5:
        return main_impl<5>(fpt_len, argc, argv);
    }
}

} // namespace

int main(int argc, char* argv[])
{
    return main_impl(get_env_uint("HAVAL_PASS", 3), get_env_uint("HAVAL_FPTLEN", 256), argc, argv);
}
